### ISSUE TITLE
Is is possible to implement a custom metric to support R^2 i.e. "Coefficient of determination" for a regression task?

### ISSUE THREAD

**Issue by aidowu1** [SciSharp/TensorFlow.NET#1222]
### Description

I need to implement a new custom metric of R-squared (coefficient of determination) for a regression problem using MLP Neural Net architecture of SciSharp's Tensorflow.Net.

I really like to use the Tensorflow.Net library for one of my ML projects and will need some help on how to correctly implement the custom metric.

Here is a simple implementation of the MLP neural net methods:

        static (NumSharp.NDArray, NumSharp.NDArray, NumSharp.NDArray, NumSharp.NDArray, NumSharp.NDArray, NumSharp.NDArray)         readAndPreprocessData()
        {
            var (X,y) = Helpers.ReadCsv(Configs.DATA_FILE_PATH, Configs.FILE_DELIMITER);
            var (X_shuffled, y_shuffled) = Helpers.shuffleDataset(X, y);
            var (X_train_valid, X_test, y_train_valid, y_test) = Helpers.splitDataset(
                X_shuffled, 
                y_shuffled, 
                split_factor: Configs.TRAIN_VALID_SPLIT_FRACTION);
            return (X, y, X_train_valid, X_test, y_train_valid, y_test);
        }

        static TensorNumpy.NDArray convertToCSharpFloatArrayFromNdarray(NumSharp.NDArray x) 
        {
            var x_array = x.ToMuliDimArray<float>();
            var x_tensor_array = TensorNumpy.np.array(x_array);
            return x_tensor_array;            
        } 

        static TensorNumpy.NDArray convertToCSharpIntArrayFromNdarray(NumSharp.NDArray x) 
        {
            var x_array = x.ToMuliDimArray<int>();
            var x_tensor_array = TensorNumpy.np.array(x_array);
            return x_tensor_array;            
        } 

        private static Tensorflow.Keras.Engine.Sequential createMLPModel(int n_features)
	    {
	        var model = KerasApi.keras.Sequential();
	        model.add(KerasApi.keras.layers.Dense(
	            units: 64,
	            input_shape: new Tensorflow.Shape(n_features),
	            activation: KerasApi.keras.activations.Relu
	        ));
	        model.add(KerasApi.keras.layers.Dense(
	            units: 64,
	            activation: KerasApi.keras.activations.Relu
	        ));
	        model.add(KerasApi.keras.layers.Dense(
	            units: 64,
	            activation: KerasApi.keras.activations.Relu
	        ));
	        model.add(KerasApi.keras.layers.Dense(
	            units: 1
	        ));
	        model.summary();
	        var metric_name = nameof(MLPNeuralNet.RSquared);
	        model.compile(
	            optimizer: KerasApi.keras.optimizers.Adam(learning_rate: Configs.LEARNING_RATE),
	            loss: KerasApi.keras.losses.MeanAbsoluteError(),
	            metrics: new[] { metric_name } // Need custom R-squared metric here
	        );
	        return model;
	    }

The entry level method call is:

		static void demoMLP()
		{
		    var (X, y, X_train_valid, X_test, y_train_valid, y_test) = readAndPreprocessData();
		    var (X_train_valid_norm, X_mean, X_std) = Helpers.normalizeData(X_train_valid);
		    var (y_train_valid_norm, y_mean, y_std) = Helpers.normalizeData(y_train_valid);
		    var n_features = X_train_valid.shape[1];
		    
		    var model = createMLPModel(n_features);
		    var X_train_valid_tensor_array = convertToCSharpFloatArrayFromNdarray(X_train_valid_norm);
		    var y_train_valid_tensor_array = convertToCSharpFloatArrayFromNdarray(y_train_valid_norm);
		    model.fit(
		        x: X_train_valid_tensor_array, 
		        y: y_train_valid_tensor_array,
		        validation_split: 0.20f,
		        epochs: 10, 
		        batch_size: 32);
		}

I also created a RSquared custom metric class, shown here: 
		
		public class RSquared : MeanMetricWrapper
		{
		    public RSquared(string name = "rSquared", TF_DataType dtype = TF_DataType.TF_FLOAT) 
		        : base((yt, yp) => Helpers.rSquared(yt, yp), 
		            name: name, 
		            dtype: dtype)
		    {
		    }
		}

The actual implementation of the R-squared logic is in the "Helpers" utility static class,  it is shown here:

    public static Tensor rSquared(Tensor y, Tensor y_hat)
    {
        var residual = tf.reduce_sum(tf.square(tf.subtract(y, y_hat)));
        var total = tf.reduce_sum(tf.square(tf.subtract(y, tf.reduce_mean(y))));
        var r2 = tf.subtract(tf.constant(1.0f), tf.divide(residual, total));
        return r2;
    }

When I build/run this code I get the following output with errors:

![image](https://github.com/SciSharp/TensorFlow.NET/assets/5054089/c7509ee5-23c0-4b80-9398-5682b594693a)

Unhandled exception. System.NotImplementedException
   at Tensorflow.Keras.Engine.MetricsContainer._get_metric_object(String metric, Tensor y_t, Tensor y_p)
   at Tensorflow.Keras.Engine.MetricsContainer.<>c__DisplayClass10_0.<_get_metric_objects>b__0(String x)
   at System.Linq.Enumerable.SelectArrayIterator`2.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Tensorflow.Keras.Engine.MetricsContainer._get_metric_objects(String[] metrics, Tensor y_t, Tensor y_p)
   at Tensorflow.Keras.Engine.MetricsContainer.Build(Tensor y_true, Tensor y_pred)
   at Tensorflow.Keras.Engine.MetricsContainer.update_state(Tensor y_true, Tensor y_pred, Tensor sample_weight)
   at Tensorflow.Keras.Engine.Model.train_step(DataHandler data_handler, Tensors x, Tensors y)
   at Tensorflow.Keras.Engine.Model.train_step_function(DataHandler data_handler, OwnedIterator iterator)
   at Tensorflow.Keras.Engine.Model.FitInternal(DataHandler data_handler, Int32 epochs, Int32 verbose, List`1 callbackList, ValidationDataPack validation_data, Func`3 train_step_func)
   at Tensorflow.Keras.Engine.Model.fit(NDArray x, NDArray y, Int32 batch_size, Int32 epochs, Int32 verbose, List`1 callbacks, Single validation_split, ValidationDataPack validation_data, Int32 validation_step, Boolean shuffle, Dictionary`2 class_weight, NDArray sample_weight, Int32 initial_epoch, Int32 max_queue_size, Int32 workers, Boolean use_multiprocessing)
   at Client.demoMLP() in C:\Development\C#\AI\Deep-Learning_Recipes\ConsoleAppBasicDemo\MLPNeuralNetConsoleApp\Program.cs:line 91
   at Client.Main(String[] args) in C:\Development\C#\AI\Deep-Learning_Recipes\ConsoleAppBasicDemo\MLPNeuralNetConsoleApp\Program.cs:line 29
